lib/recursive-methods.rb

```diff
@@ -1,49 +1,143 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n),
+#   where n is each 1 that can fit in the given number
+# Space complexity: O(n),
+#   because each n adds another method call to the stack
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError.new("#{n} is less than zero")
+  elsif n <= 0
+    return 1
+  else
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n/2),
+#   where n is the number of characters in the string,
+#   but /2 because you're chopping off two characters each time
+# Space complexity: O(n/2)
+#   because you have to add another stack frame for each n,
+#   but /2 because, again, you're chopping off two each time 
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return ""
+  elsif s.length == 1
+    return s
+  else
+    # last_char = s[-1]
+    # s = s.chop
+    # return last_char + reverse(s)
+    middle = s[1...-1]
+    return s[-1] + reverse(middle) + s[0]
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n),
+#   where n is the number of characters in the string
+# Space complexity: O(n),
+#   because you have to add a stack frame for each character
 def reverse_inplace(s)
```

Comment:
1.  This is **not** in place (see insert).
1.  You are reseting the array each time so this is O(n<sup>2</sup>)